### ISSUE TITLE
pyexe --lzma doesn't work

### ISSUE THREAD

**Issue by tansy** [upx/upx#816]
## What's the problem (or question)?

pyexe --lzma doesn't work

I tried to use upx on youtube-dl, which is pyexe under the hood. To do that I repacked it without compression and compressed with upx. While \`upx -9' seems to work, \`upx --lzma' does not.

Does not display, nor do anything.

## What should have happened?

It should work.

## Do you have an idea for a solution?

-

## How can we reproduce the issue?

To reproduce, just run it. It doesn't do anything.

```
$ ./youtube-dl --version; echo $?
2024.04.22
0

$ ./youtube-dl-0.upx4-9  --version; echo $?
2024.04.22
0

$ ./youtube-dl-0.upx4--lzma  --version; echo $?
127

$ gdb --args ./youtube-dl-0.upx4--lzma --version
Reading symbols from /root/tmp/youtube-dl-0.upx4--lzma...(no debugging symbols found)...done.
(gdb) r
Starting program: /root/tmp/youtube-dl-0.upx4--lzma --version
[Inferior 1 (process 17911) exited with code 0177]

$ strace ./youtube-dl-0.upx4--lzma --version
execve("./youtube-dl-0.upx4--lzma", ["./youtube-dl-0.upx4--lzma", "--version"], [/* 60 vars */]) = 0
getpid()                                = 17916
gettimeofday({1714439832, 681822}, NULL) = 0
unlink("/tmp/upxAUDLNHEARP2")           = -1 ENOENT (No such file or directory)
open("/tmp/upxAUDLNHEARP2", O_RDWR|O_CREAT|O_EXCL, 0700) = 3
ftruncate(3, 6023129)                   = 0
old_mmap(NULL, 6023129, PROT_READ|PROT_WRITE, MAP_SHARED, 3, 0xbfdc6e6c081bd018) = 0xb71f5000
old_mmap(0xb77b4000, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0xbfdc6e6c081bd018) = 0xb77b4000
unlink("/tmp/upxAUDLNHEARP2")           = 0
_exit(127)                              = ?
```

You can test it [here](https://files.fm/f/skqvcf3vh4), or get it from [here](https://github.com/ytdl-org/ytdl-nightly).

```
youtube-dl-2024.04.22.pyexe  ; pyexe (original)
youtube-dl-0                 ; repacked without compression
youtube-dl-0.upx4--lzma      ; upx --lzma
youtube-dl-0.upx4-9          ; upx -9
```

## Please tell us details about your environment.

* UPX version 4.2.3
* Host Operating System and version: Linux x86
* Host CPU architecture: x86
* Target Operating System and version: Linux
* Target CPU architecture: x86

**Comment by jreiser** [comment]
**How** does each particular packed version "not work"?  Copy+paste the observable symptoms (if text), or describe the video/audio problems, or (worst case) supply screenshots. State the complete command line and other necessary input that was used to cause the problems.  Running under `strace` (to capture the system calls) or `gdb` (to get a traceback at the time of a crash) will also help by giving specific details.

**Comment by tansy** [comment]
I Complemented missing data i original post.

PS. Strace of successfully running executable [looks](https://files.fm/f/qtwdktav5s) somewhat different.

**Comment by jreiser** [comment]
Fixed by commit https://github.com/upx/upx/commit/e2163d67f8efe3e08951fd2edc11e523d752a474  which currently is the tip of the main branch `devel`.  Builds are available as Artifacts on the Actions page.

**Comment by markus-oberhumer** [comment]
This commit changes the stub for i386-win32 and i386-dos32, which is not acceptable.

EDIT: I've created a new branch `jreiser-devel` and have reverted devel.

**Comment by markus-oberhumer** [comment]
@jreiser BTW, what is the status of all your public branches?

Please go to https://github.com/upx/upx/branches and delete.

**Comment by jreiser** [comment]
> This commit changes the stub for i386-win32 and i386-dos32, which is not acceptable.

@markus-oberhumer   Where is that requirement documented?  Why is not tested?  Are there any other "do not touch" cases?  Here are the related candidates:
```
$ grep -sr arch/i386/lzma_d.S stub/src
stub/src/i386-linux.kernel.vmlinuz.S:#include "arch/i386/lzma_d.S"
stub/src/i386-linux.elf.interp-entry.S:#include      "arch/i386/lzma_d.S"
stub/src/i386-win32.pe.S:#include "arch/i386/lzma_d.S"
stub/src/i386-expand.S:#include "arch/i386/lzma_d.S"
stub/src/i386-darwin.macho-entry.S:#include "arch/i386/lzma_d.S"
stub/src/i386-linux.shlib-init.S:#include "arch/i386/lzma_d.S"
stub/src/i386-bsd.elf-entry.S:#include      "arch/i386/lzma_d.S"
stub/src/i386-linux.elf.execve-entry.S:#include      "arch/i386/lzma_d.S"
stub/src/i386-dos32.djgpp2.S:#include "arch/i386/lzma_d.S"
stub/src/i386-darwin.dylib-entry.S:#include "arch/i386/lzma_d.S"
stub/src/i386-linux.kernel.vmlinux.S:#include "arch/i386/lzma_d.S"
stub/src/i386-linux.elf-entry.S:#include "arch/i386/lzma_d.S"
stub/src/i386-dos32.tmt.S:#include "arch/i386/lzma_d.S"
stub/src/i386-dos32.watcom.le.S:#include "arch/i386/lzma_d.S"
stub/src/i386-linux.elf.shell-entry.S:#include      "arch/i386/lzma_d.S"
stub/src/i386-bsd.elf.execve-entry.S:#include      "arch/i386/lzma_d.S"
```

**Comment by markus-oberhumer** [comment]
Any executable format that is possibly affected by https://github.com/upx/upx/issues/437 is only allowed real bug fixes.

**Comment by jreiser** [comment]
In `jreiser-devel` I added a new file i386/lzma-multi_d.S for Linux and BSD only; all else reverts to the old i386/lzma_d.S.
I reset my local `devel` to 8393ded1b31b93c67e970ef14af49c5524248547 which was the tip of origin/devel on  Wed May 1 13:11:12 2024 +0200, and then rebased local jreiser-devel onto local devel.  I then pushed jreiser-devel, which passed CI.
While still on jreiser-devel, then `git diff devel -- stub/i386-dos32.djgpp2.h stub/i386-dos32.tmt.h stub/i386-dos32.watcom.le.h stub/i386-win32.pe.h` showed no differences, so I believe that the win32 and dos32 stubs should be the same as before the recent changes.
I pause for approval to incorporate jreiser-devel onto devel.
